From 3ad645b9dfcc7d0f46b8f6237ba44f060c0997ee Mon Sep 17 00:00:00 2001
From: a <a>
Date: Fri, 21 Jul 2023 12:14:28 +0000
Subject: [PATCH] fixup! fixup! SERVICE: add files from qemu_src

---
 hw/misc/edu.c | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/hw/misc/edu.c b/hw/misc/edu.c
index 0cd3c55c3..9f1ca9554 100644
--- a/hw/misc/edu.c
+++ b/hw/misc/edu.c
@@ -24,7 +24,6 @@
 
 #include "qemu/osdep.h"
 #include "qemu/units.h"
-#include "qemu/cutils.h"
 #include "hw/pci/pci.h"
 #include "hw/hw.h"
 #include "hw/pci/msi.h"
@@ -268,8 +267,6 @@ static void edu_mmio_write(void *opaque, hwaddr addr, uint64_t val,
     case 0x20:
         if (val & EDU_STATUS_IRQFACT) {
             qatomic_or(&edu->status, EDU_STATUS_IRQFACT);
-            /* Order check of the COMPUTING flag after setting IRQFACT.  */
-            smp_mb__after_rmw();
         } else {
             qatomic_and(&edu->status, ~EDU_STATUS_IRQFACT);
         }
@@ -352,9 +349,6 @@ static void *edu_fact_thread(void *opaque)
         qemu_mutex_unlock(&edu->thr_mutex);
         qatomic_and(&edu->status, ~EDU_STATUS_COMPUTING);
 
-        /* Clear COMPUTING flag before checking IRQFACT.  */
-        smp_mb__after_rmw();
-
         if (qatomic_read(&edu->status) & EDU_STATUS_IRQFACT) {
             qemu_mutex_lock_iothread();
             edu_raise_irq(edu, FACT_IRQ);
@@ -386,7 +380,7 @@ static void pci_edu_realize(PCIDevice *pdev, Error **errp)
     memory_region_init_io(&edu->mmio, OBJECT(edu), &edu_mmio_ops, edu,
                     "edu-mmio", 1 * MiB);
     pci_register_bar(pdev, 0, PCI_BASE_ADDRESS_SPACE_MEMORY, &edu->mmio);
-    fprintf(stderr, "Hello QEMU from edu device\n");
+    fprintf(stderr, "Hello from QEMU Edu device!\n");
 }
 
 static void pci_edu_uninit(PCIDevice *pdev)
