From b21a47094707f3bc6c6ae4d9d94a7fff18f46905 Mon Sep 17 00:00:00 2001
From: a <a>
Date: Fri, 21 Jul 2023 14:53:52 +0300
Subject: [PATCH] SERVICE: add files from qemu_src

---
 hw/misc/edu.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/hw/misc/edu.c b/hw/misc/edu.c
index e935c418d..0f3c34707 100644
--- a/hw/misc/edu.c
+++ b/hw/misc/edu.c
@@ -267,6 +267,8 @@ static void edu_mmio_write(void *opaque, hwaddr addr, uint64_t val,
     case 0x20:
         if (val & EDU_STATUS_IRQFACT) {
             qatomic_or(&edu->status, EDU_STATUS_IRQFACT);
+            /* Order check of the COMPUTING flag after setting IRQFACT.  */
+            smp_mb__after_rmw();
         } else {
             qatomic_and(&edu->status, ~EDU_STATUS_IRQFACT);
         }
@@ -349,6 +351,9 @@ static void *edu_fact_thread(void *opaque)
         qemu_mutex_unlock(&edu->thr_mutex);
         qatomic_and(&edu->status, ~EDU_STATUS_COMPUTING);
 
+        /* Clear COMPUTING flag before checking IRQFACT.  */
+        smp_mb__after_rmw();
+
         if (qatomic_read(&edu->status) & EDU_STATUS_IRQFACT) {
             qemu_mutex_lock_iothread();
             edu_raise_irq(edu, FACT_IRQ);
@@ -380,6 +385,7 @@ static void pci_edu_realize(PCIDevice *pdev, Error **errp)
     memory_region_init_io(&edu->mmio, OBJECT(edu), &edu_mmio_ops, edu,
                     "edu-mmio", 1 * MiB);
     pci_register_bar(pdev, 0, PCI_BASE_ADDRESS_SPACE_MEMORY, &edu->mmio);
+    fprintf(stderr, "Hello QEMU from edu device\n");
 }
 
 static void pci_edu_uninit(PCIDevice *pdev)
