From 6c8068e42186ff0fe333d38ff56449989c24c0af Mon Sep 17 00:00:00 2001
From: a <a>
Date: Fri, 21 Jul 2023 12:02:43 +0000
Subject: [PATCH] fixup! SERVICE: add files from qemu_src

---
 hw/misc/edu.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hw/misc/edu.c b/hw/misc/edu.c
index 0f3c34707..0cd3c55c3 100644
--- a/hw/misc/edu.c
+++ b/hw/misc/edu.c
@@ -24,6 +24,7 @@
 
 #include "qemu/osdep.h"
 #include "qemu/units.h"
+#include "qemu/cutils.h"
 #include "hw/pci/pci.h"
 #include "hw/hw.h"
 #include "hw/pci/msi.h"
